src/test/clojure/clara/test_infinite_loops.clj

```diff
@@ -0,0 +1,189 @@
+                        (insert! (->First))]]
+
+   :sessions [empty-session [first-rule second-rule] {}]}
+
+  (let [watched-session (ld/with-loop-detection empty-session throw-fn 10)]
+
+    (assert-ex-data {:clara-rules/infinite-loop-suspected true}
+                    (-> watched-session
+                        (insert (->First))
+                        (fire-rules {:max-cycles 10})))))
+
+(def-rules-test test-recursive-insertion-loop-with-salience
+
+  {:rules [first-rule [[[First]]
+                       (insert! (->Second))
+                       {:salience 1}]
+
+           second-rule [[[Second]]
+                        (insert! (->First))
+                        {:salience 2}]]
+
+   :sessions [empty-session [first-rule second-rule] {}
+              empty-session-negated-salience [first-rule second-rule] {:activation-group-fn (comp - :salience :props)}]}
+
+  (doseq [session
+          (map #(ld/with-loop-detection % throw-fn 10)
+               [empty-session empty-session-negated-salience])]
+    (assert-ex-data {:clara-rules/infinite-loop-suspected true}
+                    (-> session
+                        (insert (->First))
+                        (fire-rules {:max-cycles 10})))))
+
+;; FIXME: Fix this before merging; it is a genuine issue to address not a test problem.
+;; (def-rules-test test-tracing-infinite-loop
+
+;;   {:rules [cold-rule [[[Cold]]
+;;                       (insert! (->Cold "ORD"))]]
+
+;;    :sessions [empty-session [cold-rule] {}]}
+
+;;   (let [watched-session (ld/with-loop-detection empty-session throw-fn 10)]
+
+;;     (try
+;;       (do (-> watched-session
+;;               tr/with-tracing
+;;               (insert (->Cold "ARN"))
+;;               ;; Use a small value here to ensure that we don't run out of memory before throwing.
+;;               ;; There is unfortunately a tradeoff where making the default number of cycles allowed
+;;               ;; high enough for some use cases will allow others to create OutOfMemory errors in cases
+;;               ;; like these but we can at least throw when there is enough memory available to hold the facts
+;;               ;; created in the loop.
+;;               (fire-rules {:max-cycles 10}))
+;;           (is false "The infinite loop did not throw an exception."))
+;;       (catch Exception e
+;;         (let [data-maps (ex-data-maps e)
+;;               loop-data (filter :clara-rules/infinite-loop-suspected data-maps)]
+;;           (is (= (count loop-data)
+;;                  1)
+;;               "There should only be one exception in the chain from infinite rules loops.")
+;;           (is (= (-> loop-data first  :listeners count) 1)
+;;               "There should only be one listener.")
+;;           (is (-> loop-data first  :listeners ^PersistentTracingListener first .-trace not-empty)
+;;               "There should be tracing data available when a traced session throws an exception on an infinite loop."))))))
```

Comment:
This is still an issue with the listener approach?
I'm not sure there necessarily should/can be a trace available when arbitrary exceptions are thrown during rule evaluation.

If we want that, I think that's a broader issue than just specifically having a trace when this max cycle logic throws.

Reply:
If there is an infinite loop I think it is useful to able to use a trace to show which rules are looping.  My approach in https://github.com/cerner/clara-rules/pull/458 allows the trace to be added.